# Here is where we will have our code for the Final Project itself. 

## Table of Contents

This cells serves as an outline of what we can do for this project. It is based on A4.

### Part 1: Load and Clean the Data
#### 1A) Loading the data
#### 1B) Finding missing data
#### 1C) Cleaning up the missing data
#### 1D) Standardize and transform the data

### Part 2: Exploratory Data Visualization
This part depends on the features of our dataset. Since our hypothesis focuses on console and year, we can plot distributions of this against global and/or regional sales.
#### 2A) Scatter Matrix
#### 2B) Bar Chart
#### 2C) Histogram

### Part 3: Exploring the Data
In this part, we will figure out basic properties of the data. We can look at the different statistical properties of the different features in the dataset. 

### Part 4: Testing Distributions
#### 4A) Determining Normality

### Part 5: Data Analysis
We proposed that we would use Multiple Linear Regression to determine the p-values of the features we are interested in, and then compare this to an alpha level of 0.001 (?). We will also use an ANOVA Test (a statistical test to figure out the differences among group means in a sample) on the categorical features of interest.

Afterwards, we could explore some more, and determine if there are any confounding variables.


### Part 6: Conclusions and Summary

### Some extra stuff we CAN do (if time permits)
We could use clustering to determine distinct clusters of the video games below. Suppose we find a cluster that achieves a relatively high number of sales. This can then lead us to believe that video games with the properties defined by that cluster are the properties that strongly correlate with increased video game sales.

### Research Question: What factors lead to the greatest accumulation of global video game sales from 1980-2017?
### Hypothesis: We believe that the console itself has the most impact on video game sales, whereas year released is the least impactful. Consoles will have the most impact on video game sales because of affordability, customer brand loyalty, and the notion that people who grow up with one console only tend to buy games compatible with that console only. The year the game was released would be least impactful because each year, a number of factors can change sales, and so the year released would be confounded with these various factors.

## Import necessary packages

In [1]:
import pandas as pd

## Part 1: Loading in the Data

In [2]:
df = pd.read_csv('vgsalesGlobale.csv')

In [3]:
df.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

In [4]:
print('No. rows: ', df.shape[0])
print('No. cols: ', df.shape[1])

No. rows:  16598
No. cols:  11


### Finding and Cleaning Missing Data

In [5]:
# Get the number of observations that are missing data for each column
no_missing = df.isnull().sum()
percent_missing = no_missing / df.shape[0]
missing_data_df = pd.DataFrame([no_missing, percent_missing]).T
missing_data_df.columns = ['No. Missing', 'Percentage Missing']

In [6]:
missing_data_df

No. Missing  Percentage Missing
Rank                  0.0            0.000000
Name                  0.0            0.000000
Platform              0.0            0.000000
Year                271.0            0.016327
Genre                 0.0            0.000000
Publisher            58.0            0.003494
NA_Sales              0.0            0.000000
EU_Sales              0.0            0.000000
JP_Sales              0.0            0.000000
Other_Sales           0.0            0.000000
Global_Sales          0.0            0.000000

Above is a dataframe showing the number of missing data for each column, along with the percentage of the data that they make up. Below, we extract these rows and place them into `missing_rows`.

In [7]:
missing_rows = df[df.isnull().any(axis=1)]
missing_rows

Rank                                               Name Platform  \
179      180                                    Madden NFL 2004      PS2   
377      378                                   FIFA Soccer 2004      PS2   
431      432                         LEGO Batman: The Videogame      Wii   
470      471                         wwe Smackdown vs. Raw 2006      PS2   
607      608                                     Space Invaders     2600   
624      625                                          Rock Band     X360   
649      650           Frogger's Adventures: Temple of the Frog      GBA   
652      653        LEGO Indiana Jones: The Original Adventures      Wii   
711      713                                     Call of Duty 3      Wii   
782      784                                          Rock Band      Wii   
1126    1128                            Call of Duty: Black Ops       PC   
1133    1135                                          Rock Band      PS3   
1303    1305                                     Triple Play 99       PS   
1433    1435                       LEGO Harry Potter: Years 5-7      Wii   
1498    1500                         LEGO Batman: The Videogame      PSP   
1513    1515                                          Adventure     2600   
1585    1587                                             Combat     2600   
1649    1651                                NASCAR Thunder 2003      PS2   
1662    1664       Shrek / Shrek 2 2-in-1 Gameboy Advance Video      GBA   
1697    1699                          Hitman 2: Silent Assassin       XB   
1837    1839                                          Rock Band      PS2   
1990    1992                        Legacy of Kain: Soul Reaver       PS   
2019    2021                               Donkey Kong Land III       GB   
2086    2088                       LEGO Harry Potter: Years 5-7     X360   
2113    2115                                     Air-Sea Battle     2600   
2114    2116                                       Suikoden III      PS2   
2134    2136                                           Yakuza 4      PS3   
2145    2147                       LEGO Harry Potter: Years 5-7      PS3   
2173    2175                                   Wheel of Fortune      PS2   
2222    2224                                 Bentley's Hackpack      GBA   
...      ...                                                ...      ...   
15606  15609         Writing and Speaking Beautiful Japanese DS       DS   
15652  15655                                       Virtua Quest      PS2   
15697  15700            Shonen Jump's Yu-Gi-Oh! GX Card Almanac       DS   
15717  15720                                    Without Warning       XB   
15739  15742                              Football Manager 2007     X360   
15788  15791  Moshi, Kono Sekai ni Kami-sama ga Iru to suru ...      PSV   
15811  15814                       Ferrari: The Race Experience      Wii   
15865  15868                  PDC World Championship Darts 2008       DS   
15876  15879                    Dinotopia: The Sunstone Odyssey       GC   
15878  15881                             Dance! It's Your Stage      Wii   
15900  15903                                        Jet Impulse       DS   
15915  15918                                       Dream Dancer       DS   
16057  16060                  PDC World Championship Darts 2008      PSP   
16058  16061                  WRC: FIA World Rally Championship       PC   
16065  16068                       Aquaman: Battle for Atlantis       XB   
16191  16194                    Homeworld Remastered Collection       PC   
16194  16197                                             Shorts       DS   
16198  16201             AKB1/48: Idol to Guam de Koishitara...     X360   
16208  16211  Super Robot Monkey Team: Game Boy Advance Vide...      GBA   
16229  16232                        Brothers in Arms: Furious 4     X360   
16246  16249                       Agarest Senki: Re-app

`missing_rows` contains all of the rows that contain at least one missing value. Of all of the features in our data, only `Year` and `Publisher` are missing. 

In [8]:
print('Proportion of total data that has at least one missing value: ', len(missing_rows) / df.shape[0])

Proportion of total data that has at least one missing value:  0.018496204361971323


As you can see, the proportion of our data where there is at least one missing value takes on a very small amount: < 2%.
Given that such a small percentage of our data is missing values, we have decided that it would be quicker and easier to simply drop these rows.

In [9]:
df.dropna(inplace=True)

In [10]:
print('New amount of rows: ', df.shape[0])
print('New amount of cols: ', df.shape[1])
df.describe()

New amount of rows:  16291
New amount of cols:  11


Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16291.000000  16291.000000  16291.000000  16291.000000  16291.000000   
mean    8290.190228   2006.405561      0.265647      0.147731      0.078833   
std     4792.654450      5.832412      0.822432      0.509303      0.311879   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4132.500000   2003.000000      0.000000      0.000000      0.000000   
50%     8292.000000   2007.000000      0.080000      0.020000      0.000000   
75%    12439.500000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2020.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16291.000000  16291.000000  
mean       0.048426      0.540910  
std        0.190083      1.567345  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.480000  
max       10.570000     82.740000

Using `df.describe()`, we can see a quick summary of our dataset. 

Here are some interesting things to note from the summary:

- The maximum `Year` is 2020, which is obviously wrong, since the current year is 2019. Below, we look more into the row(s) with this `Year`.
- The minimum `NA_Sales`, `ED_Sales`, `JP_Sales`, and `Other_Sales` is 0.0. These `Sales` are measured in the millions, so a value of 0.0 would mean that the game sold less than a million copies. However, it could also mean that the game was not sold in that particular region. We will keep track of this insight.
- The `Rank` of each game is unique. That means there are 16,291 unique ranks. These ranks result from the game's `Global_Sales`, which is the feature we are trying to predict/ make models from. So, we should ignore this feature. Below, we drop the `Rank` feature.

In [17]:
df.drop('Rank')

KeyError: "['Rank'] not found in axis"

Let's look at the game made in `2020`.

In [11]:
df.loc[df['Year'] == 2020]

Rank                    Name Platform    Year       Genre Publisher  \
5957  5959  Imagine: Makeup Artist       DS  2020.0  Simulation   Ubisoft   

      NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
5957      0.27       0.0       0.0         0.02          0.29

Since it is only one row, simply Googling this can help us fix the `Year`. After a quick search, the correct `Year` is: 2009.

In [12]:
df.loc[df['Year'] == 2020, 'Year'] = 2009
df.describe()

Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16291.000000  16291.000000  16291.000000  16291.000000  16291.000000   
mean    8290.190228   2006.404886      0.265647      0.147731      0.078833   
std     4792.654450      5.831475      0.822432      0.509303      0.311879   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4132.500000   2003.000000      0.000000      0.000000      0.000000   
50%     8292.000000   2007.000000      0.080000      0.020000      0.000000   
75%    12439.500000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2017.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16291.000000  16291.000000  
mean       0.048426      0.540910  
std        0.190083      1.567345  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.480000  
max       10.570000     82.740000

### Standardizing the Data

